R/maic_anchored.R

```diff
@@ -208,20 +216,53 @@ maic_anchored_tte <- function(res,
   # get bootstrapped estimates if applicable
   res$inferential[["boot_est"]] <- NULL
   if (!is.null(weights_object$boot)) {
-    tmp_boot_obj <- weights_object$boot
-    k <- dim(tmp_boot_obj)[3]
-    tmp_boot_est <- sapply(seq_len(k), function(ii) {
-      boot_x <- tmp_boot_obj[, , ii]
-      boot_ipd_id <- weights_object$data$USUBJID[boot_x[, 1]]
-      boot_ipd <- ipd[match(boot_ipd_id, ipd$USUBJID), , drop = FALSE]
-      boot_ipd$weights <- boot_x[, 2]
+    keep_rows <- setdiff(seq_len(nrow(weights_object$data)), weights_object$rows_with_missing)
+    boot_ipd_id <- weights_object$data[keep_rows, "USUBJID", drop = FALSE]
+
+    boot_ipd <- merge(boot_ipd_id, ipd, by = "USUBJID", all.x = TRUE)
+    if (nrow(boot_ipd) != nrow(boot_ipd_id)) stop("ipd has multiple observations for some patients")
+
+    stat_fun <- function(data, index, w_obj, pseudo_ipd) {
+      boot_ipd <- data[index, ]
+      r <- dynGet("r", ifnotfound = NA) # Get bootstrap iteration
+      if (!is.na(r)) {
+        if (isFALSE(all.equal(index, w_obj$boot[, 1, r]))) stop("Bootstrap and weight indices don't match")
+        boot_ipd$weights <- w_obj$boot[, 2, r]
+      }
+      boot_dat <- rbind(boot_ipd, pseudo_ipd)
+      boot_dat$ARM <- factor(boot_dat$ARM, levels = c(trt_agd, trt_ipd))
+      boot_coxobj_dat_adj <- coxph(Surv(TIME, EVENT) ~ ARM, boot_dat, weights = weights)
+      c(est = coef(boot_coxobj_dat_adj)[1] - coef(coxobj_agd)[1], var = vcov(boot_coxobj_dat_adj)[1, 1])
```

Comment:
From which data should the bootstrap be sampling from?  Both arms of `ipd` ? Also from `pseudo_ipd`?
Should it calculate just the A vs C model? Should it do the ACvBC comparison within the bootstrap?
```
c(est = coef(boot_coxobj_dat_adj)[1] - res_BC$est, var = vcov(boot_coxobj_dat_adj)[1, 1] + res_BC$se^2)
```
or not?
```
c(est = coef(boot_coxobj_dat_adj)[1], var = vcov(boot_coxobj_dat_adj)[1, 1])
```
Any thoughts? @hoppanda ? @MikeJSeo

Reply:
Here's what I remember:
- You would never bootstrap aggregate data as it will change the baseline covariate averages which we are adjusting our ipd matching to.
- So the bootstrap function bootstraps only the ipd trial
- So, in the stat_fun function, we shouldn't subtract by agd coefficient. What stat_fun returns is: ipd trial coef and standard error
- Then finally, we can use bucher function to find effect of BC. 
- I think we might also be interested in the effect of AC (where A: ipd experimental C: control; to be discussed)

Reply:
What I'm still not sure about is if we do many times inside the bootstrap, or just once. 
If we do it after, we need:

- the effect estimate: from the full data
- a standard error of the estimate: from the bootstrap samples

It's not clear to me if/when we could calculate the bootstrap CIs eg with BCa

Reply:
I'm not sure what your question is.
c(est = coef(boot_coxobj_dat_adj)[1] - coef(coxobj_agd)[1], var = vcov(boot_coxobj_dat_adj)[1, 1])

so in this part, we don't need to return variance. 
all we need is the point estimate. We calculate point estimate across multiple bootstraps. We calculate CI/SD once using bootstrapped point estimates.

Reply:
Ok, so we have the result from boot() with all of the point estimates from the samples. Then what exactly do you propose to calculate? Do you think we use boot.ci?

(The variance is used here for some types of boot.ci types, but we can ignore that for now.)

Reply:
Yea I think so.

You have already nicely written some code using boot.ci. I think we should take out w_obj and pseudoipd though?

boot_ci <- boot.ci(boot_res, type = boot_ci_type, w_obj = weights_object, pseudo_ipd = pseudo_ipd)

    l_u_index <- switch(boot_ci_type,
      "norm" = list(2, 3, "normal"),
      "basic" = list(4, 5, "basic"),
      "stud" = list(4, 5, "student"),
      "perc" = list(4, 5, "percent"),
      "bca" = list(4, 5, "bca")
    )

Reply:
We have:
1. point estimate of B v C from paper/pseudo IPD
2. point estimates of A v C from full data
3. bootstrap point estimates of A v C

I'm trying to figure out how you think we get to a CI for A v B.

 - Do you want the boot.ci() of 3. to get a CI for A v C? then...?
 - Or do you want to take the SE based on 3. and put it into Bucher?
 - Or something else?

Reply:
I see where your question comes from now..

I think one way to calculate A v B is just:
- bootstrap SE from (3)
- put it in to Bucher.

other way is
put bucher function inside stat_fun and return estimate of A v B there.

Reply:
From meeting, prefer to do Bucher inside the bootstrap